### PR TITLE
Add float16 support

### PR DESCRIPTION
**Pull request by duli2012**
We don't need this PR since we plan to produce a stand-along float16 converter.

### CODE REVIEW

File: onnxmltools/convert/common/_container.py [hunk 1]
```diff
@@ -128,13 +130,18 @@ def __init__(self, targeted_onnx):
         self.nodes = []
         # ONNX operators' domain-version pair set. They will be added into opset_import field in the final ONNX model.
         self.node_domain_version_pair_sets = set()
+        # Precision of float numbers: 16 or 32
```
Comment (wschin) [inline onnxmltools/convert/common/_container.py]: Please throw an error if it's neither 16 or 32. Also, do we need to support Double?

File: onnxmltools/convert/common/_container.py [hunk 2]
```diff
@@ -112,8 +113,9 @@ class ModelComponentContainer:
     encapsulated in a ONNX ModelProto.
     '''
 
-    def __init__(self, targeted_onnx):
+    def __init__(self, float_precision=32, targeted_onnx):
         '''
+        :param float_precision: precision of float number, 32 bits or 16 bits
```
Comment (wschin) [inline onnxmltools/convert/common/_container.py]: Precision of floats (or Precision of floating-point numbers), 32 bits or 16 bits. Applicable to ONNX inputs, outputs and initializers.

File: onnxmltools/convert/common/_container.py [hunk 3]
```diff
@@ -112,8 +113,9 @@ class ModelComponentContainer:
     encapsulated in a ONNX ModelProto.
     '''
 
-    def __init__(self, targeted_onnx):
+    def __init__(self, float_precision=32, targeted_onnx):
```
Comment (wschin) [inline onnxmltools/convert/common/_container.py]: this signature should be (self, targeted_onnx, float_precision=32)

File: onnxmltools/convert/common/_container.py [hunk 4]
```diff
@@ -164,6 +171,8 @@ def add_initializer(self, name, onnx_type, shape, content):
         :param shape: Tensor shape, a list of integers.
         :param content: Flattened tensor values (i.e., a float list or a float array).
         '''
+        if (self.float_precision == 16 and onnx_type == onnx_proto.TensorProto.FLOAT):
```
Comment (wschin) [inline onnxmltools/convert/common/_container.py]: Please remove "(" and ")".

File: onnxmltools/convert/common/_container.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 # --------------------------------------------------------------------------
 
 import six
+from ...proto import helper, onnx_proto
```
Comment (wschin) [inline onnxmltools/convert/common/_container.py]: from distutils.version import StrictVersion 
from ...proto import helper, onnx_proto

**Comment by wschin** [comment]
Please also convert floating-point numbers into 16 bits if a Node have float attributes. This can be done in ModelComponentContainer's add_node(…).

File: onnxmltools/convert/common/_container.py (hunk 5 above)
Comment (duli2012) [reply onnxmltools/convert/common/_container.py]: done

File: onnxmltools/convert/common/_container.py (hunk 4 above)
Comment (duli2012) [reply onnxmltools/convert/common/_container.py]: done

File: onnxmltools/convert/common/_container.py (hunk 1 above)
Comment (duli2012) [reply onnxmltools/convert/common/_container.py]: done

**Comment by wschin** [comment]
Also add tests please. Thanks!

**Comment by justinormont** [comment]
Will we support bfloat16?

https://en.wikipedia.org/wiki/Bfloat16_floating-point_format

**Comment by duli2012** [comment]
We'll support IEEE float16 in another standalone converter.

**Comment by justinormont** [comment]
@duli2012: [bfloat16](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) differs from [IEEE float16](https://en.wikipedia.org/wiki/Half-precision_floating-point_format). 

The bfloat16 is a truncated float32 which maintains the dynamic range of the float32 (~1e-38 to ~3e38.) while reducing the significand's precision. 

The gain is to avoid the overflow/underflow that a standard float16 would hit due to its limited range (~6e-8 to 65504). In a float16, an exploding gradient can very easily hit the 66k limit of a float16. This otherwise is rather tricky for a normal coder to avoid in the float16, but easily handled in the bfloat16's larger range. Tensorflow's ops normally use bfloat16 in mixed precision by doing multiplies in bfloat16 and summation into float32.  

Conversion to/from float32 to bfloat16 is quite trivial as seen in Tensorflow's [bfloat16 code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/bfloat16.cc).

The main limiting factor for why to not support bfloat16 in ONNX is the lack of current hardware support. Only Google's TPUs, and various upcoming Intel processors support bfloat16, which limits the number of models which will contain bfloat16. Whereas IEEE float16 has wide support in GPUs.

**Comment by duli2012** [comment]
Thanks for explanation, @justinormont .
Our current plan is suppport IEEE float16.

**Pull request closed without merging**